### PR TITLE
Interpolate date

### PR DESCRIPTION
**Pull request by Rlamboll**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR
Lets `interpolate` work on datetimes as well as integer years. Makes clear that `fill_series` works on both too.

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Next release
 
+## API changes
+
+PR [#333](https://github.com/IAMconsortium/pyam/pull/333) changes the second input name of `fill_series()` to `time`. It can still be an integer.
```
Comment (danielhuppmann) [inline RELEASE_NOTES.md]: This seems to be PR #334...?

File: pyam/core.py [hunk 2]
```diff
@@ -343,24 +343,29 @@ def pivot_table(self, index, columns, values='value',
...
-        if year in df.columns:
-            df = df[np.isnan(df[year])]
-        fill_values = df.apply(fill_series,
-                               raw=False, axis=1, year=year)
+        # drop time-rows where values are already defined
+        if time in df.columns:
+            df = df[np.isnan(df[time])]
+        if self.time_col == "year":
+            if not isinstance(time, int):
+                raise ValueError("the year value, {} is not an integer".format(time))
```
Comment (danielhuppmann) [inline pyam/core.py]:
```suggestion
                raise ValueError("the `time` argument `{}` is not an integer".format(time))
```

File: RELEASE_NOTES.md (hunk 1 above)
Comment (Rlamboll) [reply RELEASE_NOTES.md]: Resolved in the new commit, not sure how you can see the number before making a PR?

Comment (danielhuppmann) [reply RELEASE_NOTES.md]: you can guess it by looking at the issues and PR list (hoping that nobody creates a new issue/PR while you push. Or you first create a PR, then add another commit adding the release-notes update.

File: tests/test_timeseries.py [hunk 3]
```diff
@@ -14,6 +15,15 @@ def test_fill_series():
     assert fill_series(y, 2006) == 3.
 
 
+def test_fill_series_years():
```
Comment (danielhuppmann) [inline tests/test_timeseries.py]:
```suggestion
def test_fill_series_datetime():
```
and suggest to rename the function above to `test_fill_series_year()`

File: tests/test_core.py [hunk 4]
```diff
@@ -668,6 +668,22 @@ def test_interpolate(test_df_year):
...
+    if test_df.time_col == "year":
+        with pytest.raises(ValueError):
+            test_df.interpolate(some_date)
+    else:
+        test_df.interpolate(some_date)
+        obs = test_df.filter(time=some_date).data['value']\
+            .reset_index(drop=True)
+        exp = pd.Series([3, 1.5, 4], name='value')
+        pd.testing.assert_series_equal(obs, exp, check_less_precise=True)
+        test_df.interpolate(some_date)
```
Comment (danielhuppmann) [inline tests/test_core.py]: can you please also insert the comment from the test above (`# redo the interpolation and check that no duplicates are added`) to make it easier for the next person looking at this test to know why this is done...

File: pyam/core.py [hunk 5]
```diff
@@ -343,24 +343,31 @@ def pivot_table(self, index, columns, values='value',
...
                               values='value', aggfunc=np.sum)
-        # drop year-rows where values are already defined
-        if year in df.columns:
-            df = df[np.isnan(df[year])]
-        fill_values = df.apply(fill_series,
-                               raw=False, axis=1, year=year)
+        # drop time-rows where values are already defined
+        if time in df.columns:
+            df = df[np.isnan(df[time])]
+        if self.time_col == "year":
```
Comment (danielhuppmann) [inline pyam/core.py]: it would be more elegant to move this sanity-check as first operation of this function (before the pivot), then it is not necessary to do the if-then statement at all (I think)

```
if self.time_col == "year" and not isinstance(time, int):
    msg = 'the `time` argument `{}` is not an integer'.format(time)
    raise ValueError(msg.format(time))
```
Added side bonus - the error is raised faster (not first do the pivot, then check).

PS: also subtle change above for consistency - trying to avoid `"` in favour of `'`

File: tests/test_core.py [hunk 6]
```diff
@@ -668,6 +668,22 @@ def test_interpolate(test_df_year):
     assert not test_df_year.filter().data.duplicated().any()
 
 
+def test_interpolate_datetimes(test_df):
+    # test that interpolation also works with date-times.
+    some_date = datetime.datetime(2007, 7, 1)
+    if test_df.time_col == "year":
+        with pytest.raises(ValueError):
```
Comment (danielhuppmann) [inline tests/test_core.py]: I think this can be `pytest.raises(ValueError, test_df, interpolate, time=some_date)` for saving a line...

File: RELEASE_NOTES.md [hunk 7]
```diff
@@ -1,5 +1,13 @@
 # Next release
 
+## API changes
+
+PR [#334](https://github.com/IAMconsortium/pyam/pull/334) changes the second input name of `fill_series()` to `time`. It can still be an integer.
```
Comment (danielhuppmann) [inline RELEASE_NOTES.md]:
```suggestion
PR [#334](https://github.com/IAMconsortium/pyam/pull/334) changes the arguments of `IamDataFrame.interpolate()` and `pyam.fill_series()` to `time`. It can still be an integer (i.e., year).
```

File: pyam/core.py [hunk 8]
```diff
@@ -343,24 +343,31 @@ def pivot_table(self, index, columns, values='value',
...
                               values='value', aggfunc=np.sum)
-        # drop year-rows where values are already defined
-        if year in df.columns:
-            df = df[np.isnan(df[year])]
-        fill_values = df.apply(fill_series,
-                               raw=False, axis=1, year=year)
+        # drop time-rows where values are already defined
+        if time in df.columns:
+            df = df[np.isnan(df[time])]
+        if self.time_col == 'year':
```
Comment (danielhuppmann) [inline pyam/core.py]: this if-else statement can be removed, the actual code is the same in both cases

**Comment by danielhuppmann** [comment]
@gidden, over to you for a final look-over and merge!

**Comment by gidden** [comment]
Looks great everybody. Thanks a bunch @Rlamboll !!
